File: LeetCodePractice/src/backtracking/MatchsticksToSquare.java
package backtracking;

import java.util.Arrays;

/*
Remember the story of Little Match Girl? By now, you know exactly what matchsticks the little match girl has, please find out a way you can make one square by using up all those matchsticks. You should not break any stick, but you can link them up, and each matchstick must be used exactly one time.

Your input will be several matchsticks the girl has, represented with their stick length. Your output will either be true or false, to represent whether you could make one square using all the matchsticks the little match girl has.

Example 1:
Input: [1,1,2,2,2]
Output: true

Explanation: You can form a square with length 2, one side of the square came two sticks with length 1.
Example 2:
Input: [3,3,3,3,4]
Output: false

Explanation: You cannot find a way to form a square with all the matchsticks.
Note:
The length sum of the given matchsticks is in the range of 0 to 10^9.
The length of the given matchstick array will not exceed 15.

time complexity: O(4^N)
NP complete problem
Naive DFS for small input size
 */
public class MatchsticksToSquare {
    public boolean makesquare(int[] nums) {
        if(nums == null || nums.length < 4){
            return false;
        }
        Arrays.sort(nums);
        int sum = 0;
        for (int n : nums) {
            sum += n;
        }
        if (sum % 4 != 0) {
            return false;
        }
        int side = sum / 4;
        return dfs(nums, new int[4], side, nums.length - 1);
    }

    private boolean dfs(int[] nums, int[] arr, int side, int pos) {
        for (int n : arr) {
            if (n > side) {
                return false;
            }
        }
        if (pos == -1) {
            return true;
        }
        for (int i = 0; i < arr.length; i++) {
            if (i > 0 && arr[i] == arr[i - 1]) {
                continue;
            }
            arr[i] += nums[pos];
            if (dfs(nums, arr, side, pos - 1)) {
                return true;
            }
            arr[i] -= nums[pos];
        }
        return false;
    }

}

File: LeetCodePython/MatchsticksToSquare.py
"""
You are given an integer array matchsticks where matchsticks[i] is the length of the ith matchstick.
You want to use all the matchsticks to make one square. You should not break any stick, but you can link them up, and each matchstick must be used exactly one time.
Return true if you can make this square and false otherwise.

Example 1:
Input: matchsticks = [1,1,2,2,2]
Output: true
Explanation: You can form a square with length 2, one side of the square came two sticks with length 1.

Example 2:
Input: matchsticks = [3,3,3,3,4]
Output: false
Explanation: You cannot find a way to form a square with all the matchsticks.

Constraints:
1 <= matchsticks.length <= 15
1 <= matchsticks[i] <= 108

hint:
1 Treat the matchsticks as an array. Can we split the array into 4 equal halves?
2 Every matchstick can belong to either of the 4 sides. We don't know which one. Maybe try out all options!
3 For every matchstick, we have to try out each of the 4 options i.e. which side it can belong to. We can make use of recursion for this.
"""
from typing import List


class MatchsticksToSquare:
    def makesquare(self, matchsticks: List[int]) -> bool:
        if len(matchsticks) < 4:
            return False
        matchsticks.sort()
        total = sum(matchsticks)
        if total % 4 != 0:
            return False
        side = total // 4
        square = [0] * 4

        def dfs(idx):
            for n in square:
                if n > side:
                    return False
            if idx == -1:
                return True
            for i in range(4):
                if i > 0 and square[i] == square[i - 1]:
                    continue
                square[i] += matchsticks[idx]
                if dfs(idx - 1):
                    return True
                square[i] -= matchsticks[idx]
            return False

        return dfs(len(matchsticks) - 1)